packages/router/src/grpc/view-protocol-server/swap-by-commitment.ts

```diff
@@ -1,26 +1,34 @@
 import type { Impl } from '.';
 import { servicesCtx } from '../../ctx';
+import { watchSubscription } from './util/watch-subscription';
 
 import { SwapRecord } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
 
 import { ConnectError, Code } from '@connectrpc/connect';
 
-export const swapByCommitment: Impl['swapByCommitment'] = async (req, ctx) => {
+export const swapByCommitment: Impl['swapByCommitment'] = async (
+  { swapCommitment: findSwapCommitment, awaitDetection },
+  ctx,
+) => {
   const services = ctx.values.get(servicesCtx);
   const { indexedDb } = await services.getWalletServices();
-  const { swapCommitment } = req;
-  if (!swapCommitment)
+  if (!findSwapCommitment)
     throw new ConnectError('Missing swap commitment in request', Code.InvalidArgument);
 
-  const swap = await indexedDb.getSwapByCommitment(swapCommitment);
-  if (swap) return { swap };
+  // start subscription early to avoid race condition
+  const subscription = indexedDb.subscribe('SWAPS');
+
+  const swap =
+    (await indexedDb.getSwapByCommitment(findSwapCommitment)) ??
+    (awaitDetection &&
+      SwapRecord.fromJson(
+        await watchSubscription(subscription, update => {
+          const scannedSwap = SwapRecord.fromJson(update.value);
+          return findSwapCommitment.equals(scannedSwap.swapCommitment);
+        }).catch(),
+      ));
 
-  if (req.awaitDetection) {
-    for await (const { value: swapJson } of indexedDb.subscribe('SWAPS')) {
-      const swap = SwapRecord.fromJson(swapJson);
-      if (swap.swapCommitment?.equals(swapCommitment)) return { swap };
-    }
-  }
+  if (swap) return { swap };
```

Comment:
Hm... what's the reason for this change? To me, this feels a bit harder to read than how it was before.